### PR TITLE
Feat/registration

### PR DESCRIPTION
**Pull request by mattydono**
Sorry I thought I had included Login on a previous PR.

This pr includes:

- Login component with basic input functionality
- Registration component password and confirm password fields added
- validity check added and then refactored using useEffect

### CODE REVIEW

**Comment by Sharakai** [comment]
Can you split this into 2 PRs, please - do the prettier config first, and then another with the following 2 commits.
It's impossible for me to see the real changes, otherwise

File: src/Components/useField.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { ChangeEventHandler, useCallback, useState } from 'react';
+
+type UseFieldType<V = undefined, Element = HTMLInputElement> = [V, ChangeEventHandler<Element>];
+
+export const useField = <
```
Comment (Sharakai) [inline src/Components/useField.ts:5]: Expertly written hook! Beautiful. It's like an expert made it...

File: src/Registration/Registration.tsx [hunk 2]
```diff
@@ -14,76 +15,60 @@ type DispatchProps = {
...
+  const onAddUser = useCallback(() => {
+    const user = { email, firstName, surname, type, password, confirmPassword };
     if (user.type !== null) {
       // Typescript not registering null check -> as User
-      return this.props.addUser(user as User);
+      return props.addUser(user as User);
     }
-  };
-}
+  }, [email, firstName, surname, type, password, confirmPassword]);
```
Comment (Sharakai) [inline src/Registration/Registration.tsx]: This is missing `props.addUser`

File: src/Registration/Registration.tsx [hunk 3]
```diff
@@ -14,76 +15,60 @@ type DispatchProps = {
...
+
+  const disabled =
```
Comment (Sharakai) [inline src/Registration/Registration.tsx]: Why not use the `isValid` to determine if the button should be disabled?
The `useEffect` below needs to be a bit more complex, but then it'll align well with whether the button is enabled/disabled and reads more idiomatically.

File: src/Registration/Registration.tsx [hunk 4]
```diff
@@ -14,76 +15,60 @@ type DispatchProps = {
...
-    this.setState({ firstName: event.currentTarget.value });
-  };
-
-  private onSurnameChange: ChangeEventHandler<HTMLInputElement> = event => {
-    this.setState({ surname: event.currentTarget.value });
-  };
+const _Registration = (props: Props) => {
+  const [firstName, onFirstNameChange] = useField('');
+  const [surname, onSurnameChange] = useField('');
+  const [type, onTypeChange] = useField<UserType | null, HTMLSelectElement>(null);
```
Comment (Sharakai) [inline src/Registration/Registration.tsx:21]: This'll break if you change the `type` to something, and then select the "Select..." value. I'll leave it up to you to tell me why.

**Review by Sharakai — CHANGES_REQUESTED** [review]
Missing dep, and breakage if you re-select "Select..." type, needs to be fixed before it can be safely merged.

File: src/Components/useField.ts (hunk 1 above)
Comment (mattydono) [reply src/Components/useField.ts:5]: Thanks, sometimes I just managed to pull one out of the bag.

File: src/Registration/Registration.tsx (hunk 2 above)
Comment (mattydono) [reply src/Registration/Registration.tsx]: This has been added into deps.

File: src/Registration/Registration.tsx (hunk 3 above)
Comment (mattydono) [reply src/Registration/Registration.tsx]: I've taken your suggestion and moved the `const disabled` block into my useEffect to construct a more flexible and fleshed out validity check. I agree that this only makes sense as opposed to have 2 blocks of code to execute a single process.

File: src/Registration/Registration.tsx (hunk 4 above)
Comment (mattydono) [reply src/Registration/Registration.tsx:21]: This'll break due to the `type || ''` and the string portion of this union returning as `undefined` where (user) type can only be of type `UserType || null`. As discussed I'll create a function to take the undefined value and return it as a `null` value.

File: src/Registration/Registration.tsx (hunk 3 above)
Comment (Sharakai) [reply src/Registration/Registration.tsx]: This'll also work nicely when (if) you have async validation, like checking if the email already exists on the server as the user edits the form.

Comment (mattydono) [reply src/Registration/Registration.tsx]: I'll speak to Steve this evening about setting up a basic server to persist my `users{}` to. In the mean time I'll run this check locally but users won't survive post browser refresh.

Comment (Sharakai) [reply src/Registration/Registration.tsx]: Why not persist it in `localStorage`, for now? That's what mostly happens during local development - don't want to rely on a server being up just so you can develop locally...

Comment (mattydono) [reply src/Registration/Registration.tsx]: I haven't done this as discussed

**Review by Sharakai — APPROVED** [review]

**Pull request merged**